python/ray/util/spark/cluster_init.py

```diff
@@ -0,0 +1,630 @@
+
+    def disconnect(self):
+        if self.ray_context is not None:
+            try:
+                self.ray_context.disconnect()
+            except Exception as e:
+                # swallow exception.
+                _logger.warning(
+                    f"An error occurred while disconnecting from the ray cluster: "
+                    f"{repr(e)}"
+                )
+            self.ray_context = None
+        else:
+            _logger.warning("Already disconnected from this ray cluster.")
+
+    def shutdown(self):
+        """
+        Shutdown the ray cluster created by the `init_ray_cluster` API.
+        """
+        if not self.is_shutdown:
+            if self.ray_context is not None:
+                self.disconnect()
+            try:
+                self._cancel_background_spark_job()
+            except Exception as e:
+                # swallow exception.
+                _logger.warning(
+                    f"An error occurred while cancelling the ray cluster background spark job: "
+                    f"{repr(e)}"
+                )
+            try:
+                self.head_proc.kill()
+            except Exception as e:
+                # swallow exception.
+                _logger.warning(
+                    f"An Error occurred during shutdown of ray head node: {repr(e)}"
+                )
+            self.is_shutdown = True
+        else:
+            _logger.warning("The cluster has been shut down.")
+
+    def __enter__(self):
+        self.connect()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.shutdown()
+
+
+def _convert_ray_node_options(options):
+    return [f"--{k.replace('_', '-')}={str(v)}" for k, v in options.items()]
+
+
+@PublicAPI(stability="alpha")
+def init_ray_cluster(
+    num_worker_nodes=None,
+    total_cpus=None,
+    total_gpus=None,
+    total_heap_memory_bytes=None,
+    total_object_store_memory_bytes=None,
+    heap_to_object_store_memory_ratio=None,
+    head_options=None,
+    worker_options=None,
+    ray_temp_root_dir="/tmp",
+    ray_log_root_dir="/tmp",
+    safe_mode=True,
+):
+    """
+    Initialize a ray cluster on the spark cluster by starting a ray head node in the spark
+    application's driver side node.
+    After creating the head node, a background spark job is created that
+    generates an instance of `RayClusterOnSpark` that contains configuration for the ray cluster
+    that will run on the Spark cluster's worker nodes.
+    The returned instance can be used to connect to, disconnect from and shutdown the ray cluster.
+    This instance can also be used as a context manager (used by encapsulating operations within
+    `with init_ray_cluster(...):`). Upon entering the managed scope, the ray cluster is initiated
+    and connected to. When exiting the scope, the ray cluster is disconnected and shut down.
+
+    Args
+        num_worker_nodes: The number of spark worker nodes that the spark job will be submitted to.
```

Comment:
instead of -1, can we just use `None`?

Reply:
Sounds good.

Reply:
I think we'd better define a constant like `ray.util.spark.MAX_WORKERS`, and let user explicitly set `num_worker_nodes=MAX_WORKERS`, the `None` looks like a default value but defaultly setting to be "max workers" is risky, it will eat all resources of the cluster.